File: Leetcode Medium/199BinaryTreeRightSideView.py
#https://leetcode.com/problems/binary-tree-right-side-view/description/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        allList = []
        queue  = deque([(root, 0)])
        while len(queue) > 0:
            iterations = len(queue)
            for i in range(iterations):
                node, level = queue.popleft()
                if node == None: continue
                if len(allList) == level: allList.append([])
                allList[level].append(node.val)
                queue.append((node.left, level+1))
                queue.append((node.right, level+1))
        ret = [] 
        for List in allList:
            ret.append(List[-1])
        return ret

File: C++/Leetcode Medium/199BinaryTreeRightSideView.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include <vector>
#include <list>
using namespace std;
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if (!root) return {};
        list<TreeNode*> queue = {root};
        vector<int> res;
        while (!queue.empty()) {
            int n = queue.size();
            res.push_back(queue.front()->val);
            for (int i = 0; i < n; ++i) {
                TreeNode* curr = queue.front();
                queue.pop_front();
                if (curr->right) queue.push_back(curr->right);
                if (curr->left) queue.push_back(curr->left);
            }
        }
        return res;
    }
};